File: src/spacesimulation/Mass.java
package spacesimulation;

import spacesimulation.algebra.Point3d;
import spacesimulation.algebra.Vec;

import java.awt.*;

public class Mass extends Point3d implements Entity {
    private final Vec velocity;
    private final Vec acceleration;
    private final double mass;

    private InteractionStatus status = InteractionStatus.Movable;

    public Mass(double mass, double x, double y, double z) {
        super(x, y, z);
        if (mass == 0) throw new IllegalArgumentException("Mass can't be equal to 0");
        velocity = new Vec(0, 0, 0);
        acceleration = new Vec(0, 0, 0);
        this.mass = mass;
    }

    public Mass(double mass, Point3d pos) {
        this(mass, pos.x, pos.y, pos.z);
    }

    public Mass(double mass, Vec positionVector) {
        this(mass, positionVector.x, positionVector.y, positionVector.z);
    }

    @Override
    public void tick() {
        accelerate();
        move();
    }

    @Override
    public void render(Graphics3d drawer, Graphics g) {
        drawer.drawDot(this, 4, Color.white, g);
    }

    private void accelerate() {
        velocity.add(acceleration);
        acceleration.scale(0);
    }

    private void move() {
        add(velocity);
    }

    public void applyForce(Vec force) {
        this.acceleration.add(Vec.scale(force, 1.0 / mass));
    }

    public void accelerate(Vec acceleration) {
        this.acceleration.add(acceleration);
    }

    public void removeAccelerationInDirection(Vec direction) {
        acceleration.add(acceleration.linearProjection(direction).scale(-1));
    }

    public Vec getImpulse() {
        return Vec.scale(velocity, mass);
    }

    public Vec getCurrentForce() {
        return acceleration;
    }

    public Vec getVelocity() { return velocity; }

    public void setVelocity(Vec velocity) {
        this.velocity.x = velocity.x;
        this.velocity.y = velocity.y;
        this.velocity.z = velocity.z;
    }

    public double getMass() { return mass; }

    public Vec getAcceleration() { return acceleration; }

    public InteractionStatus getStatus() {
        return status;
    }

    public void setStatus(InteractionStatus status) {
        this.status = status;
    }

    public enum InteractionStatus {
        Immovable,
        Movable
    }
}

File: src/main/java/spacesimulation/physics/Mass.kt
package spacesimulation.physics

import spacesimulation.Entity
import spacesimulation.Graphics3d
import spacesimulation.algebra.Point3d
import spacesimulation.algebra.Vec
import java.awt.Color
import java.awt.Graphics

open class Mass(mass: Double, x: Double, y: Double, z: Double): Point3d(x, y, z), Entity {
    var velocity: Vec
    private val acceleration: Vec
    val mass: Double
    var status = Status.Movable

    init {
        require(mass != 0.0) { "Mass can't be equal to 0" }
        velocity = Vec(0.0, 0.0, 0.0)
        acceleration = Vec(0.0, 0.0, 0.0)
        this.mass = mass
    }

    constructor(mass: Double, pos: Point3d) : this(mass, pos.x, pos.y, pos.z)
    constructor(mass: Double, positionVector: Vec) : this(mass, positionVector.x, positionVector.y, positionVector.z)

    override fun tick(dt: Seconds) {
        accelerate(dt)
        move(dt)
    }

    override fun render(drawer: Graphics3d, g: Graphics) {
        drawer.drawDot(this, 0.25, Color.white, g)
    }

    private fun accelerate(dt: Seconds) {
        velocity.add(Vec.scale(acceleration, dt))
        acceleration.scale(0.0)
    }

    private fun move(dt: Seconds) {
        add(velocity * dt)
    }

    fun applyForce(force: Vec) {
        acceleration.add(force / mass)
    }

    fun accelerate(acceleration: Vec) {
        this.acceleration.add(acceleration)
    }

    fun removeAccelerationInDirection(direction: Vec) {
        acceleration.sub(acceleration.linearProjection(direction))
    }

    val impulse: Vec
        get() = Vec.scale(velocity, mass)


    enum class Status {
        Immovable,
        Movable
    }
}
